queryable-spec.ts

```diff
@@ -0,0 +1,424 @@
+   */
+  count?: 'estimate' | 'exact';
+};
+
+
+
+/**
+ * A FilterResult is an object that represents the result of a filter 
+ * expression of FilterableSource for a given quad pattern and expression. 
+ * It MAY create results lazily after one of its methods is invoked.
+ */
+interface FilterableResult {
+
+  /**
+   * Returns a Stream containing all the quads that matched the given quad
+   * pattern and expression.
+   */
+  execute(opts?: { order?: QueryOperationOrder<TermName> }): Promise<Stream<RDF.Quad>>;
+
+  /**
+   * Asynchronously returns a QueryResultMetadata, that contains the metadata
+   * of the current result.
+   */
+  metadata(opts?: FilterableResultMetadataOptions): Promise<FilterableResultMetadata>;
+  
+  /**
+   * Asynchronously returns a boolean indicating if the requested expression is
+   * supported by the FilterableSource. If it returns true, quads() and 
+   * metadata() MAY produce a valid result. If it returns false, quads() MUST
+   * return a stream emitting an error, and metadata() MUST reject.
+   */
+  isSupported(): Promise<boolean>;
+};
+
+/* 
+ * A FilterableSource is an object that produces a FilterableSourceResult that
+ * can emit quads. The emitted quads can be directly contained in this 
+ * FilterableSourceo bject, or they can be generated on the fly.
+ * 
+ * FilterableSource is not necessarily an extension of the RDF/JS Source 
+ * interface, but implementers MAY decide to implement both at the same time.
+ * 
+ * matchExpression() Returns a FilterableSourceResult that contains a quad 
+ * stream that processes all quads matching the quad pattern and the expression.
+ * 
+ * When a Term parameter is defined, and is a NamedNode, Literal or BlankNode, 
+ * it must match each produced quad, according to the Quad.equals semantics. 
+ * When a Term parameter is a Variable, or it is undefined, it acts as a 
+ * wildcard, and can match with any Term.
+ * 
+ * NOTES:
+ * - When matching with graph set to undefined or null it MUST match all the
+ *   graphs (sometimes called the union graph). To match only the default graph
+ *   set graph to a DefaultGraph.
+ * - When an Expression parameter is defined, the complete quad stream is
+ *   filtered according to this expression. When it is undefined, no filter is
+ *   applied.
+ * 
+ * If parameters of type Variable with an equal variable name are in place,
+ * then the corresponding quad components in the resulting quad stream MUST be
+ * equal.
+ * Expression's MAY contain Variable Term's. If their variable names are equal
+ * to Variable's in the given quad pattern, then the Expression MUST be 
+ * instantiated for each variable's binding in the resulting quad stream when
+ * applying the Expression filter.
+ */
+interface FilterableSource {
+  matchExpression(
+    subject?: RDF.Term,
+    predicate?: RDF.Term,
+    obj?: RDF.Term,
+    graph?: RDF.Term,
+    expression?: Expression,
+    opts?: { 
+      limit?: number; 
```

Comment:
Within SPARQL algebra, `start` and `length` are used instead of `offset` and `limit`. Since we're here closer to the algebra than the SPARQL syntax, perhaps it makes sense to align with that instead?

Reply:
No hard preference from me, although I do think offset and limit are a bit more universal than start and length. Then again, you use start and length when dealing with buffers... Happy to align.